### ISSUE TITLE
Glitch: right-click menu disappears immediately

### ISSUE THREAD

**Issue by jsmaniac** [racket/drracket#72]
The right-click menu in DrRacket sometimes disappears instantly, or sometimes appears as a small rectangle a few pixels high (ususally, it is briefly displayed correctly then turns into the tiny rectangle).

It happens on my computer (NixOs Linux 64-bits, Enlightenment 16 desktop environment), but I also saw it happen on Windows.

Here's a video showing the bug: https://vimeo.com/181820735 (zipped video here for achival purposes: [bug.mkv.zip](https://github.com/racket/drracket/files/459902/bug.mkv.zip))

With small files the bug does not occur very often (about 10% of the time), but on larger files (or maybe with more tabs open?) it occurs more frequently, and sometimes I have to do a dozen right-clicks before I can access the menu. This bug is not something new, it's been there for several months at least.

**Comment by samth** [comment]
This seems to happen much much more frequently when attempting to set breakpoints after clicking Debug.

**Comment by samth** [comment]
cc @spall

**Comment by samth** [comment]
I just tried this with `PLT_GTK2=1` and it does not seem any different.

**Comment by jsmaniac** [comment]
I never use the debugger, so that must not be the only cause. Superstition led me to think that the menu disappeared whenever there's a garbage collection immediately after it is opened.

I tried right-clicking on `x` in the simple program below with a program running in a second tab:

```
#lang racket
(define x "hello world")
(displayln x)
```

I tried these two programs in the second tab. The first stresses the GC a lot, while the second just computes fxxor in a loop. With the `thrash1` program running, the right-click menu disappears 4 times out of 5, roughly, but with the `thrash2` program, the right-click menu never disappears. @samth can you test this and see if you get the same results?

```
#lang racket
(define (thrash1 l)
  (thrash1 (reverse l)))
(thrash1 (range 1000000))
```

```
#lang racket
(require racket/fixnum)
(define (thrash2 x)
  (thrash2 (fxxor x 1234)))
(thrash2 9999)
```

**Comment by samth** [comment]
I see basically the same behavior as you with the check-syntax right click menus. For the debugger "add breakpoint" menu it isn't any better even with nothing else running.

**Comment by AlexKnauth** [comment]
This happens on @stchang's lab desktop as well. @stchang, do you have any good cases where it definitely happens or definitely doesn't happen?

(Everyone else, it seems to happen almost always on his computer)

**Comment by stchang** [comment]
Yes it happens all the time and it's been that way for awhile . I think it's a Linux issue. @asumu reports seeing it frequently as well I believe.

**Comment by stchang** [comment]
Oops i meant @takikawa

**Comment by jsmaniac** [comment]
I experimented with this issue a bit, and tried to add `(sleep 2)` at the end of the right-click handler at https://github.com/racket/gui/blob/master/gui-lib/framework/private/keymap-global.rkt#L53-L76 :+1: 

```
 (when (send event button-up?)
   …
   (sleep 2))
```

The menu shows up correctly, stays for two seconds, and then, most of the time, either disappears, or gets collapsed to a tiny rectangle (the first "glitch" case in the video here: https://vimeo.com/181820735).

Therefore, something wrong is happening *after* the menu is built, and it seems less likely that this is a multi-thread issue, as the menu can be displayed fine, but is destroyed afterwards.

I also noticed that when I open a large file, during background compilation, the menu sometimes shows up correctly for a few seconds (and I can interact with it by hovering the menu items), but then disappears after a random amount of time (between 2 to 7 seconds in my case). Quite often, immediately right-clicking again gives a different menu (e.g. the "add require prefix" entry got added). Is it possible that when more information is gathered by the background expansion, something re-initialises menu-related stuff in some way or another? I'm not sure if the two problems are related, though (menu disappearing when new information is available, and menu disappearing immediately, because the second case also occurs even if background expansion already fully completed a long time ago).

**Comment by jsmaniac** [comment]
Oh, I found the cause of this issue. The problem is that the pop-up menu is created and displayed with `(make-object popup-menu%)`, but it seems that this object is not stored anywhere.

When the next garbage collection occurs, the menu object is discarded.

To solve this issue for the editor right-click menu, I simply had to declare a global variable just above `keymap:setup-global` in https://github.com/racket/gui/blob/master/gui-lib/framework/private/keymap-global.rkt#L45 :

```
(define global-popup-menu #f)
(define keymap:setup-global
  …)
```

Then, after the object is created, i store it in the global variable to ensure that it is not discarded by the GC on [line 56](https://github.com/racket/gui/blob/master/gui-lib/framework/private/keymap-global.rkt#L56):

```
                  (let ([m (make-object popup-menu%)])
                    (set! global-popup-menu m)
                    …)
```

The proper fix is probably to make sure that the administrator obtained with `(send edit get-admin)` saves the `popup-menu%` somewhere when `(send a popup-menu m (+ x 1) (+ y 1))` is called, until it disappears.
